audio/stereo_panner.rs

```diff
@@ -0,0 +1,139 @@
+use block::{Block, Chunk, Tick, FRAMES_PER_BLOCK};
+use euclid::Vector3D;
+use node::{AudioNodeEngine, AudioNodeMessage, BlockInfo};
+use node::{AudioNodeType, ChannelInfo};
+use param::{Param, ParamDir, ParamType};
+use std::f32::consts::PI;
+use std::cmp::min;
+use std::cmp::max;
+
+// .normalize(), but it takes into account zero vectors
+use panner_node::normalize_zero;
+
+#[derive(Copy, Clone, Debug)]
+pub struct StereoPannerOptions {
+    pub pan: f32,
+}
+
+pub enum StereoPannerNodeMessage {
+    SetPan(f32),
+}
+
+impl Default for StereoPannerOptions {
+    fn default() -> Self {
+        StereoPannerOptions {
+            pan: 0.,
+        }
+    }
+}
+
+#[derive(AudioNodeCommon)]
+pub(crate) struct StereoPannerNode {
+    pan: Param,
+}
+
+impl StereoPannerNode {
+    pub fn new(options: StereoPannerOptions, channel_info: ChannelInfo) -> Self {
+        Self {
+            pan: options.pan,
+        }
+    }
+
+    pub fn update_parameters(&mut self, info: &BlockInfo, tick: Tick) -> bool {
+        self.pan.update(info, tick)
+    }
+}
+
+impl AudioNodeEngine for StereoPannerNode {
+    fn node_type(&self) -> AudioNodeType {
+        AudioNodeType::StereoPannerNode
+    }
+
+    fn process(&mut self, mut inputs: Chunk, info: &BlockInfo) -> Chunk {
+        debug_assert!(inputs.len() == 1);
+
+        return inputs;
+
+        {
+            let block = &mut inputs.blocks[0];
+
+            block.explicit_repeat();
+
+            let mono = if block.chan_count() == 1 {
+                block.resize_silence(2);
+                true
+            } else {
+                debug_assert!(block.chan_count() == 2);
+                false
+            };
+
+                let (l, r) = block.data_mut().split_at_mut(FRAMES_PER_BLOCK.0 as usize);
+                for frame in 0..FRAMES_PER_BLOCK.0 {
+                    let frame = Tick(frame);
+                    self.update_parameters(info, frame);
+                    let mut pan = self.pan;
+
+                    // https://webaudio.github.io/web-audio-api/#stereopanner-algorithm
+
+                    // clamp pan to [-1, 1]
+                    pan = max(-1, pan);
+                    pan = min(1, pan);
+
+                    let x = if mono {
+                        x = (pan + 1) / 2;
+                    } else if pan <= 0. {
+                        x = pan + 1;
```

Comment:
almost all of the numerical values here need to have a decimal point after them to make them into floats